### ISSUE TITLE
Payment Processed Despite Failure to Record Order in Database

### ISSUE THREAD

**Issue by macserv** [concrete5-community-store/community_store#488]
I recently ran into a situation where my schema was out of sync, and any order placed by a customer was not being successfully recorded in the database.  The customer received an error, but their card was charged.  In most instances, they tried again, and received the same result, causing double-charges that had to be refunded.  Since the order was not recorded, I only had the L4 of their card to aid in identifying them, and resolving the charges was made quite difficult.

The database error itself (shown below) was my fault, and that is not the subject of this report.  The real issue here is that the customer's card was charged, despite the fact that no record of the order had been entered into the database.  In this case, the card should not have been processed.

A better flow would be to add the order provisionally to the database in an "unpaid" state, and validate that this has occurred successfully; at that point,  attempt to process the payment.

**C5 Version**: 8.5.1
**Community Store Version**: 2.1.7
**Payment Method**: Square (community_store_square 0.9.1)

**Sample database error from the C5 Log**
```
Exception Occurred: <rootPathRedacted>/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDrive An exception occurred while executing 'INSERT INTO CommunityStoreOrders (cID, oDate, pmID, pmName, smName, sInstructions, sShipmentID, sRateID, sCarrier, sTrackingID, sTrackingCode, sTrackingURL, oShippingTotal, oTax, oTaxIncluded, oTaxName, oTotal, transactionReference, oPaid, oPaidByUID, oCancelled, oCancelledByUID, oRefunded, oRefundedByUID, oRefundReason, externalPaymentRequested, locale, userAgent) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [0, "<oDateRedacted>", 2, "<pmNameRedacted>", "<smNameRedacted>", null, null, null, null, null, null, null, <oShippingTotalRedacted>, "", "", "", <oTotalRedacted>, null, null, null, null, null, null, null, null, null, "<localeRedacted>", null]:

SQLSTATE[42S22]: Column not found: 1054 Unknown column 'userAgent' in 'field list' (0)
```

**Comment by Mesuva** [comment]
You make a good point here. This might be able to be achieved, we'd just have to address the potential issue of lots of incomplete order records being added for failed payments, and then that does things like increment the order number unusually. 

I can't help but think though that this is a failing in the gateway to record enough information about the customer/order for you to then track them down. Several other gateways record the customer's name and email - I don't know much about what Square records as I've not used it myself on a project.

But let me think on this, I do like the safety of recording the order initially.

**Comment by losttheplot** [comment]
I too had a similar set of embarrassing circumstances to macserv with a previous incarnation of community_store, which resulted in our client having to piece together order and payment details from the gateway records. The cause of the issue was database related (our problem), but the most embarrassing thing was that we couldn't sort it all out for the client, because we didn't have access to their gateway account payments data section. This meant that we had to inform the client of the problem with no idea of its full extent. This was mega-embarrassing!

I write my own order processing modules for our hotel and motorsport booking systems, and the idea that macserv describes is how we've done things for the last 20 years. All the customer and order details are recorded in the Orders table prior to the gateway call, but flagged with a status of 'about to pay'. At this point, an email is dispatched to the operator with a subject of 'about to pay'.

Once the gateway callback is received, the bank's transaction details are added to the order record and the status is changed to 'paid', 'unsuccessful', 'aborted' or whatever is most appropriate. A second email is then dispatched to the operator accordingly.

This means that the operator gets used to receiving two emails in quick succession for each order: an 'about to pay' and a 'paid'. If they only get one, something has gone wrong, in which case the operator has all the details ready to immediately phone the customer and enquire if they need any hep with their order e.g. lost connection, forgotten 3D secure password, bank decides it doesn't like the weather today, whatever. Because the customer is still in a 'buying' frame of mind, this means that most failed payment attempts can be rescued within minutes.

The fact that some order numbers relate to unsuccessful payments has never been an issue, but is simple to workaround if you feel the need. Simply have an Order ID which is the table's primary key, and an Order Number that is used for administrative processing. The Order Number can be added as an incremental action upon a successful callback. This is how we always handle invoice generation because: (1) we use server and db clustering, so record primary keys are always offset by three or four (which has always caused us issues when installing concrete5 with it's hard-coded primary site records), and (2) the UK taxman likes to see invoice numbers run incrementally else he starts looking for nonexistent foul-play.

In short: I support macserv's issue or feature request, however you want to look at it.

**Comment by Mesuva** [comment]
I will take another look at this at some stage.
It does come down a bit to what payment gateway is in use. For example, the new Stripe payment method, the one that uses Checkout, by the nature of how it works always records the order before handling the payment step. The same goes for Paypal, and a few others that go off-set for the payment processing.

It's the on-site payment processors which there is a potential problem, which I'll look to address.

**Comment by Mesuva** [comment]
I think at this point in time I'll probably not adjust the way this is handled - with the move for many payment gateways to being externally processed (Stripe Checkout being the main one that comes to mind), orders are created before payments anyway.
